### PR TITLE
Add AWS integration as another strategy

### PR DESCRIPTION
**Pull request by mpass99**
Make Poseidon able to use more execution strategies. Specifically, add an integration for AWS Lambda functions as executor.

ToDo:
- Wait for #91 to be fixed :heavy_check_mark: 
- Store the usedRunners. Runner AWS Manager Get, Return. :heavy_check_mark:
- User Tests :heavy_check_mark:
- Unit Tests :heavy_check_mark:
- E2e Tests :heavy_check_mark:
- File Deletion (another PR?!) :heavy_check_mark:
- AWS Environment Manager Statistics route (another PR?!)
- Integrate authentication with AWS (Another PR)
- Dynamically create Execution environments (another PR!)

### CODE REVIEW

**Review by MrSerth — COMMENTED** [review]
That's a huge work 💪 including many refactorings 💅 
It took me some while to dive into these many changes and while I was able to understand most parts (at least I hope so), I marked very few lines with some specific questions. Other places just fix some typos or have minor suggestions, nothing too wild.

Regarding the other three open todos: I would prefer to have a dedicated PR for these (maybe targeting this PR?) so that they are easier to review 😁

File: internal/runner/aws_runner.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	err = wsConn.WriteMessage(websocket.TextMessage, rawData)
+	if err != nil {
+		exit <- ExitInfo{uint8(1), fmt.Errorf("cannot send aws request: %w", err)}
+		return
+	}
+
+	exitCode, err := w.receiveOutput(wsConn, stdout, stderr, ctx)
+	if w.TimeoutPassed() {
+		err = ErrorRunnerInactivityTimeout
+	}
```
[the remark this answers is not in the document]
Comment (mpass99) [reply internal/runner/aws_runner.go:142]: With AWS all data is dropped. We have no way of terminating an AWS Function anyway (Currently - still thinking of switching to the reverse-shell-way some day). The AWS Lambda function is terminated after the global AWS timeout:
https://github.com/openHPI/poseidon/blob/ec64a3a0969ff4e69e3708b5973c41e2082a0869/deploy/aws/template.yaml#L9-L11

For Nomad, we persist most of the timeout information. When we claim a runner we save in Nomad the information of a runner being used and with which timeout:
https://github.com/openHPI/poseidon/blob/ec64a3a0969ff4e69e3708b5973c41e2082a0869/internal/runner/nomad_manager.go#L69
The `ResetTimeout` functionality is not synced to Nomad. This is compensated for by the fact that when Poseidon is started, the timeout of all used runners is reset to the beginning.

File: internal/runner/aws_runner_test.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		exit, _, err := r.ExecuteInteractively(defaultExecutionID, nil, io.Discard, io.Discard)
+		require.NoError(t, err)
+		<-exit
+		assert.True(t, awsMock.hasConnected)
+	})
+
+	t.Run("sends execution request", func(t *testing.T) {
+		awsMock.ctx, cancel = context.WithTimeout(context.Background(), tests.ShortTimeout)
+		defer cancel()
+		command := "sl"
```
[the remark this answers is not in the document]
Comment (mpass99) [reply internal/runner/aws_runner_test.go:85]: :steam_locomotive:

File: internal/runner/aws_runner_test.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+	var cancel context.CancelFunc
+	awsMock := &awsEndpointMock{}
+	s := httptest.NewServer(http.HandlerFunc(awsMock.handler))
+
+	// Convert http://127.0.0.1 to ws://127.0.0.
+	config.Config.AWS.Endpoint = "ws" + strings.TrimPrefix(s.URL, "http")
+	awsMock.ctx, cancel = context.WithTimeout(context.Background(), tests.ShortTimeout)
+	defer cancel()
+	command := "sl"
```
[the remark this answers is not in the document]
Comment (mpass99) [reply internal/runner/aws_runner_test.go:114]: Unfortunately, only a placeholder that is doomed to never come to real execution in this unit test :cold_sweat:

File: internal/runner/manager.go [hunk 4]
```diff
@@ -67,6 +51,12 @@ type ExecutionEnvironment interface {
 // Manager keeps track of the used and unused runners of all execution environments in order to provide unused
 // runners to new clients and ensure no runner is used twice.
```
[the remark this answers is not in the document]
Comment (mpass99) [reply internal/runner/manager.go:52]: Hm, but both the specific manager and the abstract manager implement this interface. We don't say how they keep track, just that they do - and both manager implementations do so. The one by itself the other by using the first..
Are you missing a comment here?

**Review by mpass99 — COMMENTED** [review]
It' s nice to get reviews from you! :)
I agree having dedicated PRs for the open ToDos but I would prefer to not let them target this PR as I do not intend to solve all of them in the near future..

File: internal/runner/aws_runner.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	go w.handleRunnerTimeout(ctx, exitInternal, exit, execution.ID(id))
+
+	return exit, cancel, nil
+}
+
+// UpdateFileSystem copies Files into the executor.
+func (w *AWSFunctionWorkload) UpdateFileSystem(request *dto.UpdateFileSystemRequest) error {
+	for _, path := range request.Delete {
+		delete(w.fs, path)
+	}
```
[the remark this answers is not in the document]
Comment (MrSerth) [reply internal/runner/aws_runner.go:94]: Let's add a quick comment to these lines above and let's collect this limitation for future work.

File: configuration.example.yaml [hunk 6]
```diff
@@ -41,6 +43,14 @@ nomad:
   # Nomad namespace to use. If unset, 'default' is used
   namespace: poseidon
 
+aws:
+  # Specifies whether AWS should be used as executor.
+  enabled: false
+  # The enpoint of the WebSocket API
+  endpoint: wss://abcdef1234.execute-api.eu-central-1.amazonaws.com/production
+  # Currently, only static AWS environments are supported. You can list them here.
+  functions: "java11Exec go118Exec"
```
[the remark this answers is not in the document]
Comment (MrSerth) [reply configuration.example.yaml:52]: Let's extract this to a new issue and just clarify the comment ("... separated by spaces") aboove

File: internal/runner/aws_runner.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrSerth) [reply internal/runner/aws_runner.go:142]: Shall we document these behaviors somewhere?

**Review by MrSerth — APPROVED** [review]
Thanks for integrating my review comments and creating the three tickets 🎫 

I think, everything is done by now and I don't see further issues we need to tackle. 🙌

**Pull request merged**